Fix encoding array values.

mpranjic opened the pull request:
Fixes #28.

I haven't run integration tests.
Please double-check :)

sorentwo commented on lib/customer.ex:
```diff
@@ -85,7 +85,7 @@ defmodule Braintree.Customer do
       :ok = Braintree.Customer.delete("customer_id")
   """
-  @spec delete(binary) :: :ok | :error
+  @spec delete(binary) :: :ok | {:error, Error.t}
```
Thanks for fixing this.

sorentwo commented on lib/subscription.ex:
```diff
@@ -177,4 +177,18 @@ defmodule Braintree.Subscription do
+  @doc """
+  Convert a map into a Subscription struct. Add_ons and transactions
+  are converted to a list of structs as well.
+  ## Example
```
Please add space above and below the `## Example` line.

sorentwo commented on lib/transaction.ex:
```diff
@@ -225,4 +225,23 @@ defmodule Braintree.Transaction do
+      transaction = Braintree.Transaction.construct(%{"subscription_id" => "subxid",
+                                                      "status" => "submitted_for_settlement"})
+  """
+  def construct(map_or_array) do
```
I believe I'll need to restructure how `construct` gets used so that these overrides aren't necessary. In the meantime, can you please define this as two functions with guards, rather than as a single function with a case.

sorentwo commented on lib/xml/encoder.ex:
```diff
@@ -26,23 +26,22 @@ defmodule Braintree.XML.Encoder do
 
   defp generate(term) when is_map(term),
-    do: term |> Enum.into([]) |> generate
+    do: term |> Map.to_list() |> Enum.map(&generate/1) |> Enum.join("\n")
```
Nice changes here. This whole thing can be a bit mind bending.

mpranjic commented:
Thanks for the comments, I've updated the PR.
(I squashed the changes)

sorentwo commented:
Several of the integration tests are failing, which I believe is related to the xml encoding changes. You can see an isolated case of this in `test/integration/customer_test.exs`. Can you take a look?

mpranjic commented:
Sure, I'll take a look tomorrow.

mpranjic commented:
Few tests were failing because we were passing a keyword list instead of a map.
`%{...}` was missing.

I've fixed those.
I had only one test failing, but it also happens in **master**, can you confirm?
(I added the fix as separate commit)

```
test/integration/customer_test.exs:59
Assertion with =~ failed
code: error.message() =~ ~r"cvv is required"i
lhs:  "Do Not Honor"
rhs:  ~r/cvv is required/i
```

Anyway, I'd just like to add `add_ons` to the subscription tests tomorrow.

mpranjic commented:
@sorentwo I've added a few related changes.

When removing add_ons from subscription, you need to pass a list of add_on ids. [source](https://developers.braintreepayments.com/reference/request/subscription/create/ruby#add_ons.remove)
This was raising an exception because `generate` was missing an implementation for Strings.

Changes:
- modified a Subscription.create test to include the add_ons with all 3 actions (add, remove, update).
- added a test for encoding lists of strings.
- documented a sandbox environment setup needed for tests to pass.

sorentwo commented:
Great set of changes. I've modified my sandbox account with the add-ons as outlined in the new documentation. I'm going to check it out locally, make sure everything passes, and squash a few commits. Nice work 👍

sorentwo commented:
Merged into master. I'll work on getting a release out today.

mpranjic commented:
Thank you :)
